File: letterRanking.py
import requests
from collections import Counter
from typing import List, Dict

# 単語リストをキャッシュするためのグローバル変数
_word_list: List[str] = []
WORD_LIST_URL = 'https://gist.githubusercontent.com/dracos/dd0668f281e685bad51479e5acaadb93/raw/6bfa15d263d6d5b63840a8e5b64e04b382fdb079/valid-wordle-words.txt'

def get_word_list() -> List[str]:
    """
    Wordleの単語リストを取得し、キャッシュします。
    2回目以降の呼び出しではキャッシュされたリストを返します。
    """
    global _word_list
    if _word_list:
        return _word_list

    try:
        response = requests.get(WORD_LIST_URL)
        response.raise_for_status()  # HTTPエラーがあれば例外を発生させる
        
        # テキストを改行で分割し、各単語の前後の空白を削除し、5文字の単語のみをフィルタリング
        words = [
            word.strip() for word in response.text.splitlines() 
            if len(word.strip()) == 5
        ]
        _word_list = words
        return _word_list
    except requests.exceptions.RequestException as e:
        print(f"単語リストの取得に失敗しました: {e}")
        return []

def _rank_letters(words: List[str], letters_to_check: str) -> List[str]:
    """
    指定された文字セット（母音または子音）の出現頻度を計算し、
    ランキング順にソートされたリストを返します。
    """
    letter_counts = Counter()
    for word in words:
        # 単語内のユニークな文字のみをカウント対象とする
        unique_letters_in_word = set(char for char in word if char in letters_to_check)
        letter_counts.update(unique_letters_in_word)
    
    # 出現頻度が高い順に文字をソート
    # Counter.most_common() は (要素, カウント) のタプルリストを返す
    sorted_letters = [letter for letter, count in letter_counts.most_common()]
    return sorted_letters

def consonant_ranking(words: List[str]) -> List[str]:
    """子音の出現頻度ランキングを返します。"""
    consonants = "bcdfghjklmnpqrstvwxyz"
    return _rank_letters(words, consonants)

def vowel_ranking(words: List[str]) -> List[str]:
    """母音の出現頻度ランキングを返します。"""
    vowels = "aeiou"
    return _rank_letters(words, vowels)

def get_letter_ranking() -> Dict[str, List[str]]:
    """
    単語リストを取得し、子音と母音の出現頻度ランキングを計算します。
    """
    words = get_word_list()
    if not words:
        return {"consonants": [], "vowels": []}
        
    consonants = consonant_ranking(words)
    print('Consonant Ranking:', consonants)
    
    vowels = vowel_ranking(words)
    print('Vowel Ranking:', vowels)

    return {"consonants": consonants, "vowels": vowels}

if __name__ == '__main__':
    # このファイルが直接実行された場合にのみ、以下の処理を行う
    get_letter_ranking()

File: letterRanking.js
// 'https://gist.githubusercontent.com/dracos/dd0668f281e685bad51479e5acaadb93/raw/6bfa15d263d6d5b63840a8e5b64e04b382fdb079/valid-wordle-words.txt'

export const CONSONANT_SCORES = {
    's': 21, 'r': 20, 'l': 19, 't': 18, 'n': 17, 'd': 16, 'y': 15, 'm': 14, 'p': 13, 'c': 12,
    'h': 11, 'g': 10, 'b': 9, 'k': 8, 'w': 7, 'f': 6, 'v': 5, 'z': 4, 'j': 3, 'x': 2, 'q': 1
};

export const VOWEL_SCORES = {
    'e': 5, 'a': 4, 'o': 3, 'i': 2, 'u': 1
};

// #2
// 母音の出現頻度をカウント
// 同じ単語内に複数回登場する際は1回とする
// 昇順
export function sortByUniqueVowel(words, ascending = true) {
    const vowelCount = {};
    const vowels = 'aeiou';
    words.forEach(word => {
        const uniqueVowels = new Set(word.split('').filter(char => vowels.includes(char)));
        uniqueVowels.forEach(vowel => {
            vowelCount[vowel] = (vowelCount[vowel] || 0) + 1;
        });
    });
    // 出現頻度でソート
    const sortedVowels = Object.entries(vowelCount)
        .sort((a, b) => a[1] - b[1])
        .map(entry => entry[0]);
    if (ascending) {
        return sortedVowels;
    }
    sortedVowels.reverse();
    return sortedVowels;
}

// #2
// 母音スコアの合計でソート
export function sortWordsByVowelScore(words, ascending = true) {
    return words.slice().sort((a, b) => {
        const scoreA = a.split('').reduce((sum, char) => sum + (VOWEL_SCORES[char] || 0), 0);
        const scoreB = b.split('').reduce((sum, char) => sum + (VOWEL_SCORES[char] || 0), 0);
        if (ascending) {
            return scoreA - scoreB; // 昇順
        }
        return -(scoreB - scoreA); // 昇順
    });
}

// #3
// 子音スコアの合計でソート
export function sortWordsByConsonantScore(words, ascending = false) {
    return words.slice().sort((a, b) => {
        const scoreA = a.split('').reduce((sum, char) => sum + (CONSONANT_SCORES[char] || 0), 0);
        const scoreB = b.split('').reduce((sum, char) => sum + (CONSONANT_SCORES[char] || 0), 0);
        if (ascending) {
            return scoreA - scoreB; // 降順
        }
        return scoreB - scoreA; // 降順
    });
}

// #4
// 子音の出現頻度をカウント
// 同じ単語内に複数回登場する際は1回とする
// 降順
export function sortByConsonantCount(words, ascending = false) {
    const consonantCount = {};
    const consonants = 'bcdfghjklmnpqrstvwxyz';
    words.forEach(word => {
        const uniqueConsonants = new Set(word.split('').filter(char => consonants.includes(char)));
        uniqueConsonants.forEach(char => {
            consonantCount[char] = (consonantCount[char] || 0) + 1;
        });
    });
    // 出現頻度でソート
    const sortedConsonants = Object.entries(consonantCount)
        .sort((a, b) => b[1] - a[1])
        .map(entry => entry[0]);
        if (ascending) {
        return sortedConsonants;
    }
    sortedConsonants.reverse();
    return sortedConsonants;
}
